### PR TITLE
[Feat] Query

### PR DESCRIPTION
**Pull request by IHEII**
### Task Description

- Support Query in Go client.

### Solution Description

- Implement Query with ObTableAsyncQueryRequest.

### Passed Regressions

Pass all tests.

### Upgrade Compatibility

N.A.

### CODE REVIEW

File: protocol/border_flag.go [hunk 1]
```diff
@@ -18,3 +18,93 @@
 package protocol
 
 type ObBorderFlag int8
+
+const (
+	INCLUSIVE_START       ObBorderFlag = 0x1
```
Comment (namewxt1220) [inline protocol/border_flag.go:23]: 小驼峰命名

File: protocol/operation_type.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ * See the Mulan PSL v2 for more details.
+ * #L%
+ */
+
+package protocol
+
+type ObQueryOperationType int32
+
+const (
+	QUERY_START ObQueryOperationType = iota
```
Comment (namewxt1220) [inline protocol/operation_type.go:23]: 小驼峰命名

File: protocol/border_flag.go [hunk 3]
```diff
@@ -18,3 +18,93 @@
...
+// NewObBorderFlag creates a new ObBorderFlag.
+func NewObBorderFlag() ObBorderFlag {
+	borderFlag := ObBorderFlag(0)
+	borderFlag.SetInclusiveStart()
+	borderFlag.SetInclusiveEnd()
+	return borderFlag
+}
+
+// SetInclusiveStart sets the INCLUSIVE_START flag.
+func (bf *ObBorderFlag) SetInclusiveStart() {
```
Comment (namewxt1220) [inline protocol/border_flag.go:38]: *ObBorderFlag推荐为ObBorderFlag，因为ObBorderFlag本身就是uint8，无需设置成指针类型

File: protocol/border_flag.go [hunk 4]
```diff
@@ -18,3 +18,93 @@
...
+	*bf &= ^INCLUSIVE_END
+}
+
+// IsInclusiveEnd returns true if the INCLUSIVE_END flag is set.
+func (bf ObBorderFlag) IsInclusiveEnd() bool {
+	return bf&INCLUSIVE_END == INCLUSIVE_END
+}
+
+// SetMinValue sets the BORDER_FLAG_MIN_VALUE flag.
+func (bf ObBorderFlag) SetMinValue() {
```
Comment (namewxt1220) [inline protocol/border_flag.go:68]: set unset会改变bf原本值，但bf并非为指针类型，因此每次进行set操作时，会将bf值进行拷贝，函数内修改bf值，不会在函数外面生效，需要将bf修改为指针类型

File: protocol/border_flag.go [hunk 5]
```diff
@@ -18,3 +18,93 @@
...
+	bf |= BORDER_FLAG_MIN_VALUE
+}
+
+// UnSetMinValue unsets the BORDER_FLAG_MIN_VALUE flag.
+func (bf ObBorderFlag) UnSetMinValue() {
```
Comment (namewxt1220) [inline protocol/border_flag.go:73]: 同上

File: protocol/border_flag.go [hunk 6]
```diff
@@ -18,3 +18,93 @@
...
+	return bf&BORDER_FLAG_MIN_VALUE == BORDER_FLAG_MIN_VALUE
+}
+
+// SetMaxValue sets the BORDER_FLAG_MAX_VALUE flag.
+func (bf ObBorderFlag) SetMaxValue() {
+	bf |= BORDER_FLAG_MAX_VALUE
+}
+
+// UnSetMaxValue unsets the BORDER_FLAG_MAX_VALUE flag.
+func (bf ObBorderFlag) UnSetMaxValue() {
```
Comment (namewxt1220) [inline protocol/border_flag.go:88]: 这里同上，将bf修改为指针类型，这里建议将ObBorderFlag所有的set unset操作全部修改为指针类型，否则执行set unset操作后bf值并不会在函数外部生效

File: protocol/border_flag.go (hunk 1 above)
Comment (IHEII) [reply protocol/border_flag.go]: Fixed, thanks to @namewxt1220 for his code :D

File: protocol/operation_type.go (hunk 2 above)
Comment (IHEII) [reply protocol/operation_type.go:23]: Fixed

File: protocol/border_flag.go (hunk 4 above)
Comment (IHEII) [reply protocol/border_flag.go]: Fixed

File: protocol/border_flag.go (hunk 5 above)
Comment (IHEII) [reply protocol/border_flag.go]: Fixed.

File: protocol/border_flag.go (hunk 6 above)
Comment (IHEII) [reply protocol/border_flag.go]: Fixed.

File: client/obclient.go [hunk 7]
```diff
@@ -742,6 +763,37 @@ func (c *obClient) getPartitionId(entry *route.ObTableEntry, rowKeyValue []*tabl
...
+	if !entry.IsPartitionTable() || entry.PartitionInfo().Level() == route.PartLevelZero {
+		return []uint64{0}, nil
+	}
+	if entry.PartitionInfo().Level() == route.PartLevelOne {
+		return entry.PartitionInfo().FirstPartDesc().GetPartIds(rowKeyPair)
+	}
+	if entry.PartitionInfo().Level() == route.PartLevelTwo {
+		partIds1, err := entry.PartitionInfo().FirstPartDesc().GetPartIds(rowKeyPair)
+		if err != nil {
+			return []uint64{route.ObInvalidPartId}, errors.WithMessagef(err, "get part id from first part desc, firstDesc:%s",
```
Comment (WeiXinChan) [inline client/obclient.go:777]: return nil，error

File: client/obclient.go [hunk 8]
```diff
@@ -742,6 +763,37 @@ func (c *obClient) getPartitionId(entry *route.ObTableEntry, rowKeyValue []*tabl
...
+				entry.PartitionInfo().FirstPartDesc().String())
+		}
+		partIds2, err := entry.PartitionInfo().SubPartDesc().GetPartIds(rowKeyPair)
+		if err != nil {
+			return []uint64{route.ObInvalidPartId}, errors.WithMessagef(err, "get part id from sub part desc, firstDesc:%s",
```
Comment (WeiXinChan) [inline client/obclient.go:782]: return nil, error

File: client/obclient.go [hunk 9]
```diff
@@ -742,6 +763,37 @@ func (c *obClient) getPartitionId(entry *route.ObTableEntry, rowKeyValue []*tabl
...
+		// do cartesian product
+		partIds := make([]uint64, 0, len(partIds1)*len(partIds2))
+		for _, partId1 := range partIds1 {
+			for _, partId2 := range partIds2 {
+				partIds = append(partIds, (partId1)<<route.ObPartIdShift|partId2|route.ObMask)
+			}
+		}
+		return partIds, nil
+	}
+	return []uint64{route.ObInvalidPartId}, errors.Errorf("unknown partition level, partInfo:%s", entry.PartitionInfo().String())
```
Comment (WeiXinChan) [inline client/obclient.go:794]: return nil, error

File: client/option.go [hunk 10]
```diff
@@ -38,6 +40,43 @@ func (f ObkvOptionFunc) apply(opts *ObkvOptions) {
...
+		maxResultSize: -1,
+		limit:         -1,
+		offset:        0,
+		scanOrder:     table.Forward,
+		isHbaseQuery:  false,
+	}
+}
+
+type ObkvQueryOptions struct {
+	queryFilter   interface{}
```
Comment (WeiXinChan) [inline client/option.go:64]: why queryFilter and hTableFilter are interface{}

File: client/query.go [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+// AddKeyRanges adds key ranges.
+func (q *ObQueryExecutor) AddKeyRanges(keyRanges []*table.RangePair) {
+	if q.keyRanges == nil {
+		q.keyRanges = make([]*table.RangePair, 0, len(keyRanges))
+	}
+	q.keyRanges = append(q.keyRanges, keyRanges...)
+}
+
+// SetEntityTypes sets the entity type.
+func (q *ObQueryExecutor) SetEntityTypes(entityType protocol.ObTableEntityType) {
```
Comment (WeiXinChan) [inline client/query.go:66]: SetEntityTypes->SetEntityType

File: client/query.go [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+	}
+
+	// get table params
+	expectant, err := q.getTableParams(ctx, q.tableName, q.keyRanges, false)
+	if err != nil {
+		return nil, errors.WithMessage(err, "get table params")
+	}
+
+	// set query range into table query
+	if nil != q.transferQueryRange() {
```
Comment (WeiXinChan) [inline client/query.go:216]: err := = q.transferQueryRange()
if err != nil {
xxx
}

File: client/query_result_iterator.go [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+package client
+
+import (
+	"context"
+	"github.com/oceanbase/obkv-table-client-go/protocol"
+	"github.com/pkg/errors"
+	"sync"
+)
+
+type ObQueryResultIterator struct {
```
Comment (WeiXinChan) [inline client/query_result_iterator.go:27]: expose an interface to user

File: client/client.go [hunk 14]
```diff
@@ -84,6 +84,8 @@ type Client interface {
 	Delete(ctx context.Context, tableName string, rowKey []*table.Column, opts ...ObkvOption) (int64, error)
 	// Get a record by rowKey.
 	Get(ctx context.Context, tableName string, rowKey []*table.Column, getColumns []string, opts ...ObkvOption) (SingleResult, error)
+	// Query records by rangePairs.
+	Query(ctx context.Context, tableName string, rangePairs []*table.RangePair, opts ...ObkvQueryOption) (*ObQueryResultIterator, error)
```
[the remark this answers is not in the document]
Comment (IHEII) [reply client/client.go]: Add interface QueryResultIterator.

File: client/obclient.go (hunk 8 above)
Comment (IHEII) [reply client/obclient.go]: Fixed.

File: client/obclient.go (hunk 9 above)
Comment (IHEII) [reply client/obclient.go]: Fixed.

File: client/option.go (hunk 10 above)
Comment (IHEII) [reply client/option.go:64]: queryFilter and hTableFilter will be implemented later.

File: client/query.go (hunk 11 above)
Comment (IHEII) [reply client/query.go]: Fixed.

File: client/query.go (hunk 12 above)
Comment (IHEII) [reply client/query.go]: Fixed.

File: client/query_result_iterator.go (hunk 13 above)
Comment (IHEII) [reply client/query_result_iterator.go:33]: Add interface QueryResultIterator.

File: client/query.go [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+		return nil, errors.WithMessage(err, "get table params")
+	}
+
+	// set query range into table query
+	err = q.transferQueryRange()
+	if err != nil {
+		return nil, errors.WithMessage(err, "transfer query range")
+	}
+
+	return newObQueryResultIteratorWithParams(&ctx, q.cli, q.tableQuery, expectant, q.entityType, q.tableName), nil
```
Comment (namewxt1220) [inline client/query.go:221]: context 不要传递其指针，参数类型为context.Context接口类型，这里的ctx本身就是指针

File: client/query_result_iterator.go [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+	readConsistency       protocol.ObTableConsistencyLevel
+	tableName             string
+	prevSessionId         int64
+	rowIndex              int
+	closed                bool
+	hasNext               bool
+}
+
+// newObQueryResultWithParam creates a new ObQueryResultIterator.
+func newObQueryResultIteratorWithParams(ctx *context.Context,
```
Comment (namewxt1220) [inline client/query_result_iterator.go:51]: context 不要传递其指针，参数类型为context.Context接口类型，这里的ctx本身就是指针

File: client/query.go [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+	tableName string,
+	keyRanges []*table.RangePair,
+	refresh bool) ([]*ObTableParam, error) {
+	entry, err := q.cli.getOrRefreshTableEntry(ctx, tableName, refresh, false)
+	if err != nil {
+		return nil, errors.WithMessagef(err, "get or refresh table entry, tableName:%s", tableName)
+	}
+
+	// get partition ids from key ranges
+	partIdList := make([]uint64, 0)
```
Comment (namewxt1220) [inline client/query.go:101]: partIdList  make操作能否将容量一并初始化？这里看起来并不能得到list最终的容量大小，是否将keyRanges的长度作为其容量大小以减少扩容slice扩容次数

File: client/query.go (hunk 15 above)
Comment (IHEII) [reply client/query.go:221]: Fixed.

File: client/query.go (hunk 17 above)
Comment (IHEII) [reply client/query.go:101]: ``partIdList`` is the list of all partition Ids calculated from the query ranges, which might be hard to estimate.

File: client/query.go [hunk 18]
```diff
@@ -0,0 +1,222 @@
...
+		}
+		tableParams = append(tableParams, NewObTableParam(t, entry.TableId(), partId))
+	}
+
+	return tableParams, nil
+}
+
+// removeDuplicates removes duplicates id in partIdList.
+func removeDuplicates(nums []uint64) []uint64 {
+	set := make(map[uint64]bool)
```
[the remark this answers is not in the document]
Comment (IHEII) [reply client/query.go:137]: ``removeDuplicates()`` is used to remove the duplicate element from ``partIdList``, when there is a key partition table or a hash partition table, ``partIdList`` may contain many duplicate partitionId/tabletId calculated from query ranges provided by users.

**Comment by WeiXinChan** [comment]
LGTM

**Comment by namewxt1220** [comment]
LGTM

**Pull request merged**